book/src/developers/designs/unixd_multi_resolver_2024.md

```diff
@@ -0,0 +1,306 @@
+authentication mechanisms like CTAP2 or PIV.
+
+Unixd already needs to parse these files to understand and prevent conflicts between local items and
+remote ones. Extending this functionality will allow us to resolve local users from memory.
+
+#### Local Group Extension
+
+An often requested feature is the ability to extend a local group with the members from a remote
+group. Often people attempt to achieve this by "overloading" a group remotely such as creating a
+group called "wheel" in Kanidm and then attempting to resolve it on their systems. This can
+introduce issues as different distributions may have the same groups but with different gidNumbers
+which can break systems, or it can cause locally configured items to be masked.
+
+Instead, we should allow group _extension_. A local group can be nominated for extension, and paired
+to a remote group. For example this could be configured as:
+
+```
+[group."wheel"]
+extend_from = "opensuse_wheel"
+```
+
+This allows the local group "wheel" to be resolved and _extended_ with the members from the remote
+group `opensuse_wheel`.
+
+#### Multiple Resolvers
+
+We would like to support multiple backends simultaneously and in our source tree. This is a major
+motivator of this rework as the himmelblau project wishes to contribute their client layer into our
+source tree, while maintaining the bulk of their authentication code in a separate libhimmelblau
+project.
+
+We also want to support LDAP and other resolvers too.
+
+The major challenge here is that this shift the cache state from the daemon to the client. This
+requires each client to track it's own online/offline state and to self-handle that state machine
+correctly. Since we won't allow dynamic modules this mitigates the risk as we can audit all the
+source of interfaces committed to the project for correct behaviour here.
+
+#### Resolvers That Can't Resolve Without Authentication Attempts
+
+Some resolvers are unable to resolve accounts without actually attempting an authentication attempt
+such as Himmelblau. This isn't a limitation of Himmelblau, but of Azure AD itself.
+
+This has consequences on how we performance authentication flows generally.
+
+#### Domain Joining of Resolvers
+
+Some Clients (and in the future Kanidm) need to be able to persist some state related to Domain
+Joining, where the client registers to the authentication provider. This provides extra
+functionality beyond the scope of this document, but a domain join work flow needs to be possible
+for the providers in some manner.
+
+#### Encrypted Caches
+
+To protect caches from offline modification content should be optionally encrypted / signed in the
+future.
+
+### Challenges
+
+- The order of backend resolvers needs to be stable.
+- Accounts/Groups should _not_ move/flip-flop between resolvers.
+- Resolvers need to uniquely identify entries in some manner.
+- The ability to store persistent/permananent accounts in the DB that can _not_ be purged in a cache
```

Comment:
We should identify the purpose of this, and how we would indicate this to the end user, because I know I'd forget about it and be confused why it wouldn't go away when I did a cache clear.

Reply:
I think the context here is "that way we can store things like passkeys to a local user account". You wouldn't want those purged after a cache clear.

Anything from remote services is always cached, but we need to make sure local-things can't be accidentally stomped berause we can't just re-populate them.